substratevm/src/com.oracle.svm.core.posix/src/com/oracle/svm/core/posix/PosixJavaNetSubstitutions.java

```diff
@@ -119,84 +419,1339 @@ protected int peekData(DatagramPacket p) throws IOException {
+            //   266                          "Socket closed");
+            throw new SocketException("Socket closed");
+            //   267          return;
+        }
+        //   269      fd = (*env)->GetIntField(env, fdObj, IO_fd_fdID);
+        fd = PosixUtils.getFD(fdObj);
+        //   270
+        //   271      if (IS_NULL(address)) {
+        if (address == null) {
+            //   272          JNU_ThrowNullPointerException(env, "address");
+            throw new NullPointerException("address");
+            //   273          return;
+        }
+
+        //   276      if (NET_InetAddressToSockaddr(env, address, port, (struct sockaddr *)&rmtaddr, &len, JNI_TRUE) != 0) {
+        if (JavaNetNetUtilMD.NET_InetAddressToSockaddr(address, port, rmtaddr, len_Pointer, true) != 0)  {
+            //   277        return;
+            return;
+        }
+
+        //   280      setDefaultScopeID(env, (struct sockaddr *)&rmtaddr);
+        JavaNetNetUtilMD.setDefaultScopeID(rmtaddr);
+        //   281
+        //   282      if (JVM_Connect(fd, (struct sockaddr *)&rmtaddr, len) == -1) {
+        if (VmPrimsJVM.JVM_Connect(fd, rmtaddr, len_Pointer.read()) == -1) {
+            //   283          NET_ThrowByNameWithLastError(env, JNU_JAVANETPKG "ConnectException",
+            //   284                          "Connect failed");
+            throw new ConnectException(PosixUtils.lastErrorString("Connect failed"));
+            //   285          return;
+        }
     }
+    // @formatter:on
 
 }
 
 @TargetClass(java.net.DatagramSocket.class)
 @Platforms({Platform.LINUX.class, Platform.DARWIN.class})
 final class Target_java_net_DatagramSocket {
 
-    @Substitute
-    @SuppressWarnings("unused")
-    Target_java_net_DatagramSocket(SocketAddress bindaddr) throws SocketException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.DatagramSocket.<init>(SocketAddress)");
-    }
-
     @Substitute
     private void checkOldImpl() {
         // it calls java.net.PlainDatagramSocketImpl.peekData(DatagramPacket) which is unimplemented
     }
 
-    @Substitute
-    @SuppressWarnings({"static-method", "unused"})
-    public void bind(SocketAddress addr) throws SocketException {
-        throw VMError.unsupportedFeature("Unimplemented: java.net.DatagramSocket.bind(SocketAddress)");
+}
+
+@TargetClass(java.net.DatagramPacket.class)
+@Platforms({Platform.LINUX.class, Platform.DARWIN.class})
+final class Target_java_net_DatagramPacket {
+    /* Aliases to get visibility for substituted methods. */
+    @Alias byte[] buf;
+    @Alias InetAddress address;
+    @Alias int offset;
+    @Alias int length;
+    @Alias int bufLength;
+    @Alias int port;
+}
+
+class Util_java_net_DatagramPacket {
+    static Target_java_net_DatagramPacket as_Target_java_net_DatagramPacket(DatagramPacket packet) {
+        return KnownIntrinsics.unsafeCast(packet, Target_java_net_DatagramPacket.class);
```

Comment:
A little late, and not in the style of the rest of the code in this file, but casts between "real" Java types and substitute "Target" types can be done with `.class.cast`.  See `JavaThreads.fromTarget(Target_java_lang_Thread)` and `JavaThreads.toTarget(Thread)`.  The types really are the same at runtime, otherwise the `unsafeCast` would not work.  But you have to put the `@SuppressFBWarnings(value = "BC", justification = "Cast for @TargetClass") annotation on the casting methods or FindBugs objects that the cast can never work.